### PR TITLE
Add worker_task_slots_available metric and worker_type tag

### LINKED ISSUE #986: Add metrics for active executors threads
**Is your feature request related to a problem? Please describe.**

Right now metrics that are provided by JavaSDK don't give a direct answer to a question of what percentage of WTF, Activity, or Local Activity executors threads are actually used. 
We have a metric that gives us time between SCHEDULE and PULL of the task (SCHEDULE_TO_START) that can be judged that there are not enough Pollers.
We have a metric that gives us information on how long does it take to process a task (execution_latency)
But we don't have any direct stats helping to understand that we don't have enough executors. We can theoretically have an underloaded host with not enough executor threads and no metric to understand that we utilize the executor pool fully and we should increase its size.

**Describe the solution you'd like**
Report a gauge metric with a number of available threads in Executor ThreadPools.

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Add worker_task_slots_available metric providing information about the count of available executor threads and worker_type tag for worker-level metrics.

Closes #986

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/internal/worker/PollTaskExecutor.java [hunk 1]
```diff
@@ -91,6 +100,8 @@ public void process(T task) {
             MDC.remove(LoggerTag.NAMESPACE);
             MDC.remove(LoggerTag.TASK_QUEUE);
           }
+          // this thread will be released right after, so updating count with +1 adjuster
+          publishSlotsMetric(1);
```
Comment (cretz) [inline temporal-sdk/src/main/java/io/temporal/internal/worker/PollTaskExecutor.java:104]: I don't believe this is a thread-safe approach. If there are 10 slots and 5 being used and all 5 finish at the same time calling this, the value might end up at 10 - 5 + 1 racily, which is 6 when really 10 will be available. If you don't want to maintain a counter that you incr/decr, is there somewhere _after_ the task executes that you refresh the number?

Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/worker/PollTaskExecutor.java:104]: It's not a thread-safe approach at all. It's a good enough approach. See "Additional context" here:  https://github.com/temporalio/sdk-java/issues/998
> is there somewhere after the task executes that you refresh the number?
no, there is none right now.

File: temporal-sdk/src/main/java/io/temporal/internal/worker/PollTaskExecutor.java [hunk 2]
```diff
@@ -125,4 +136,10 @@ public String toString() {
     // that will simplify such kind of logging through workers.
     return String.format("PollTaskExecutor{name=%s, identity=%s}", pollThreadNamePrefix, identity);
   }
+
+  private void publishSlotsMetric(int adjuster) {
+    this.metricsScope
+        .gauge(MetricsType.WORKER_TASK_SLOTS_AVAILABLE)
+        .update(this.workerTaskSlots - taskExecutor.getActiveCount() + adjuster);
```
[the remark this answers is not in the document]
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/worker/PollTaskExecutor.java:143]: > How approximate?
"Good enough" approximate. It's just a metric. In real life, it definitely could miss by a thread or two or once in a while by a lot. It's "eventually consistent" though, which makes it "good enough" for a metric.
It's a temporary solution and a "right" thread-safe solution based on the semaphore is delayed till #998

[the remark this answers is not in the document]
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/worker/PollTaskExecutor.java:143]: > I wonder if keeping a counter that starts at the slot count and +1/-1'ing that atomically is safer.

I don't see a big difference for a metric honestly. But if you find it preferable, I will reimplement it this way, it's a tiny change.

**Pull request merged**